src/document_basic.c

```diff
@@ -192,26 +192,42 @@ int Document_ReplyAllFields(RedisModuleCtx *ctx, RedisModuleString *id) {
     goto done;
   }
 
-  size_t len = RedisModule_CallReplyLength(rep);
-  RS_LOG_ASSERT(len % 2 == 0, "Number of elements must be even");
+  size_t hashLen = RedisModule_CallReplyLength(rep);
+  RS_LOG_ASSERT(hashLen % 2 == 0, "Number of elements must be even");
   // Zero means the document does not exist in redis
-  if (len == 0) {
+  if (hashLen == 0) {
     RedisModule_ReplyWithArray(ctx, 0);
     goto done;
   }
 
-  size_t elen;
+  size_t strLen;
   RedisModuleCallReply *e;
-  RedisModule_ReplyWithArray(ctx, len);
-  for (size_t i = 0; i < len; ++i) {
+  SchemaRule *rule = spec->rule;
+  RedisModule_ReplyWithArray(ctx, REDISMODULE_POSTPONED_ARRAY_LEN);
+  size_t numElems = 0;
+  for (size_t i = 0; i < hashLen; i += 2) {
+    // parse field
     e = RedisModule_CallReplyArrayElement(rep, i);
-    const char *str = RedisModule_CallReplyStringPtr(e, &elen);
-    if (elen != 0) {
-      RedisModule_ReplyWithStringBuffer(ctx, str, elen);
+    const char *str = RedisModule_CallReplyStringPtr(e, &strLen);
+    RS_LOG_ASSERT(strLen > 0, "field string cannot be empty");
+    if ((rule->lang_field && strncasecmp(str, rule->lang_field, strlen(rule->lang_field)) == 0) ||
```

Comment:
`strcasecmp` isn't good enough here?

Reply:
@rafie 
No. `RedisModule_CallReplyStringPtr` does not return the usual null-terminated string.

Reply:
Ok, so can we use `strLen` instead of `strlen(...)` as the field names are null-terminated?

Reply:
No, since strLen is the length of the reply which is longer.

Reply:
We don't know which string is longer. If we reach the null-terminator then we stop, and we'll not be scanning more than n chars on either string. A simple test can verify this. I'm trying to save some cycles here.

Reply:
if the reply string is shorter it is defiantly not equal to the attribute string

Reply:
`str` is not null-terminated, thus we must include `strLen` as a stop condition to `strncasecmp` of it will not stop at the end of `str` if the fields are longer.

Reply:
So we actually need to go:
`(rule->lang_field && strlen(rule->lang_field) == strLen && !strncasecmp(str, rule->lang_field, strLen))`
with the possible optimization of doing the strlen(...) outside the loop.